### PR TITLE
imp(e2e): added generic grpc querier

### PR DESCRIPTION
**Pull request by srdtrk**
## Description

closes: #5714

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

## Summary by CodeRabbit

- **Refactor**
	- Enhanced query handling across various test suites by centralizing query operations through the `query` package, improving code maintainability and consistency.
	- Streamlined inter-blockchain communication (IBC) operations with updated query functions, enhancing efficiency in retrieving essential IBC data.
- **Tests**
	- Updated end-to-end test suites to utilize the new query structure, ensuring compatibility and robustness of the testing framework.

### CODE REVIEW

File: e2e/testsuite/query/queries.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+	}
+
+	clientStateAny := clientStateResp.ClientState
+
+	clientState, err := clienttypes.UnpackClientState(clientStateAny)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientState, nil
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply e2e/testsuite/query/queries.go]: good catch

File: e2e/testsuite/query/grpc_query.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	"strings"
+
+	"github.com/cosmos/gogoproto/proto"
+	"github.com/strangelove-ventures/interchaintest/v8/ibc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+// Queries the chain with a query request and deserializes the response to T
+func GRPCQuery[T any](ctx context.Context, chain ibc.Chain, req proto.Message, opts ...grpc.CallOption) (*T, error) {
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply e2e/testsuite/query/grpc_query.go:15]: There are cases we want to test the error. In fact, we do this to check whether a packet commitment is missing here: https://github.com/cosmos/ibc-go/pull/5979/files#diff-a36fcd03b468866255ec8560b99276d8897eccd88d3876b1534e6c9e9b48c800R347

File: e2e/testsuite/query/grpc_query.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	path, err := getProtoPath(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a connection to the gRPC server.
+	grpcConn, err := grpc.Dial(
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply e2e/testsuite/query/grpc_query.go:22]: Yeah, if there is a convincing argument showing that dialing costs matter, then I'd be happy to create this global

**Review by chatton — APPROVED** [review]
LGTM! Looks great @srdtrk, nice improvement! Just a few small comments but overall looks great!

File: e2e/testsuite/query/grpc_query.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (srdtrk) [reply e2e/testsuite/query/grpc_query.go:15]: liner complaining when I do this + I think keeping it simple is good imo

File: e2e/testsuite/query/grpc_query.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/cosmos/gogoproto/proto"
+	"github.com/strangelove-ventures/interchaintest/v8/ibc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+// Queries the chain with a query request and deserializes the response to T
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply e2e/testsuite/query/grpc_query.go]: done

File: e2e/tests/core/02-client/client_test.go [hunk 5]
```diff
@@ -118,26 +105,35 @@ func (s *ClientTestSuite) TestScheduleIBCUpgrade_Succeeds() {
...
 		// checks there is an upgraded client state stored
-		cs, err := s.QueryUpgradedClientState(ctx, chainA, ibctesting.FirstClientID)
+		upgradedCsResp, err := query.GRPCQuery[clienttypes.QueryUpgradedClientStateResponse](ctx, chainA, &clienttypes.QueryUpgradedClientStateRequest{})
+		s.Require().NoError(err)
+
+		clientStateAny := upgradedCsResp.UpgradedClientState
+
+		cfg := chainA.Config().EncodingConfig
+		var cs ibcexported.ClientState
+		err = cfg.InterfaceRegistry.UnpackAny(clientStateAny, &cs)
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply e2e/tests/core/02-client/client_test.go:115]: I've only added wrappers for queries that are repeated more than once. This is the only use of this query

[the remark this answers is not in the document]
Comment (crodriguezvega) [reply e2e/tests/core/02-client/client_test.go:116]: all right, sounds good. We can add a helper function later on if the query is used in more than one place.

**Review by crodriguezvega — APPROVED** [review]

**Pull request merged**